File: 0047-permutations-ii/0047-permutations-ii.java
/* METHOD-1: Used Array */
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    public void helper(int[] nums, boolean[] used, List<Integer> curr) {
        if (nums.length == curr.size()) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            
            if (i > 0 && nums[i] == nums[i-1] && !used[i-1]) continue;

            used[i] = true;
            curr.add(nums[i]);
            helper(nums, used, curr);
            curr.remove(curr.size() - 1);
            used[i] = false;
        } 
    }
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        boolean[] used = new boolean[nums.length];
        Arrays.fill(used, false);
        helper(nums, used, new ArrayList<>());
        return ans;
    }
}
/* METHOD-2: Swap Based */
// class Solution {
//     public void Swap(int[] nums, int i, int j) {
//         int temp = nums[i];
//         nums[i] = nums[j];
//         nums[j] = temp;
//     }

//     public List<List<Integer>> permuteUnique(int[] nums) {
        
//     }
// }

File: 0047-permutations-ii/0047-permutations-ii.cpp
/* METHOD-1: Used Array Based */
// class Solution {
// public:
//     vector<vector<int>> ans;
//     void helper(vector<int>& nums, vector<int>& curr, vector<int>& used) {
//         if (curr.size() == nums.size()) {
//             ans.push_back(curr);
//             return;
//         }
//         for (int i = 0; i < nums.size(); i++) {
//             if (used[i]) continue;

//             if (i > 0 && nums[i] == nums[i-1] && !used[i-1]) continue;

//             used[i] = 1;
//             curr.push_back(nums[i]);
//             helper(nums, curr, used);
//             curr.pop_back();
//             used[i] = 0;
//         }
//     }
//     vector<vector<int>> permuteUnique(vector<int>& nums) {
//         sort(nums.begin(), nums.end());
//         vector<int> curr;
//         vector<int> used(nums.size(), 0);
//         helper(nums, curr, used);
//         return ans;
//     }
// };

/* METHOD-2: Swap Based */
class Solution {
public:
    vector<vector<int>> ans;
    void helper(int index, vector<int>& nums) {
        if (index == nums.size()) {
            ans.push_back(nums);
            return;
        }

        unordered_set<int> seen;

        for (int i = index; i < nums.size(); i++) {
            if (seen.count(nums[i])) continue;
            seen.insert(nums[i]);

            swap(nums[index], nums[i]);
            helper(index + 1, nums);
            swap(nums[index], nums[i]);
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        helper(0, nums);
        return ans;
    }
};